File: solutions/ruby/palindrome-products/6/palindrome_products.rb
require 'set'

class Palindromes
  @@cache = {}

  def initialize(max_factor:, min_factor: 1)
    raise ArgumentError, "min must be <= max" if min_factor > max_factor
    @max_factor = max_factor
    @min_factor = min_factor
  end

  def generate
    key = [@min_factor, @max_factor]
    if @@cache[key]
      @palindromes = @@cache[key]
    else
      @palindromes = []
      palindromes_hash = {}
      (@min_factor..@max_factor).each do |i|
        (i..@max_factor).each do |j|
          product = i * j
          if palindrome?(product)
            factors_to_add = [[i, j]]
            if palindromes_hash[product]
              existing = palindromes_hash[product]
              factors_to_add.each do |factor_pair|
                existing.add_factor(factor_pair)
              end
            else
              @palindromes << PalindromeProduct.new(product, factors_to_add)
              palindromes_hash[product] = @palindromes.last
            end
          end
        end
      end
      @@cache[key] = @palindromes
    end
  end

  def smallest
    @palindromes.min_by(&:value) || PalindromeProduct.new(nil, [])
  end

  def largest
    @palindromes.max_by(&:value) || PalindromeProduct.new(nil, [])
  end
  
  def palindrome?(number)
    original = number
    reversed = 0
    while number > 0
      reversed = reversed * 10 + number % 10
      number /= 10
    end
    original == reversed
  end
end

class PalindromeProduct
  attr_reader :value

  def initialize(value, factors = [])
    @value = value
    @factors = Set.new(factors)
  end

  def factors
    @factors.to_a.sort
  end

  def add_factor(pair)
    @factors << pair
  end
end

File: solutions/csharp/palindrome-products/6/PalindromeProducts.cs
using System.Collections.Concurrent;

using Factors = System.Collections.Generic.IEnumerable<(int, int)>;

public static class PalindromeProducts
{
    public static (int, Factors) Largest(int minFactor, int maxFactor)
    {
        var palindromeProducts = GetProducts(minFactor, maxFactor);

        var largest = palindromeProducts.Keys.Max();
        return (largest, palindromeProducts[largest]);
    }

    private static Dictionary<int, Factors> GetProducts(int minFactor, int maxFactor)
    {
        var palindromeProducts = new Dictionary<int, Factors>();
        for (var i = minFactor; i <= maxFactor; i++)
        {
            for (var j = i; j <= maxFactor; j++)
            {
                if (IsPalindrome(i * j))
                {
                    var currentFactors = palindromeProducts.GetValueOrDefault(i * j, []);
                    currentFactors = currentFactors.Append((i, j));
                    palindromeProducts[i * j] = currentFactors;
                }
            }
        }

        return palindromeProducts.Count > 0 ? palindromeProducts : throw new ArgumentException();
    }

    private static bool IsPalindrome(int product)
    {
        var productString = product.ToString();
        var digits = productString.ToCharArray();
        Array.Reverse(digits);

        return new string(digits) == productString;

    }
    public static (int, Factors) Smallest(int minFactor, int maxFactor)
    {
        var palindromeProducts = GetProducts(minFactor, maxFactor);

        var smallest = palindromeProducts.Keys.Min();
        return (smallest, palindromeProducts[smallest]);
    }
}
